src/lib/viewers/controls/page/PageControlsForm.tsx

```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import { decodeKeydown } from '../../../util';
+import './PageControlsForm.scss';
+
+export type Props = {
+    onPageSubmit: (newPageNumber: number) => void;
+    pageNumber: number;
+    pageCount: number;
+};
+
+export const ENTER = 'Enter';
+export const ESCAPE = 'Escape';
+
+export default function PageControlsForm({ onPageSubmit, pageNumber, pageCount }: Props): JSX.Element {
+    const [isInputShown, setIsInputShown] = React.useState(false);
+    const [inputValue, setInputValue] = React.useState(pageNumber.toString());
+    const [isButtonFocus, setIsButtonFocus] = React.useState(false);
+    const buttonElRef = React.useRef<HTMLButtonElement>(null);
+    const inputElRef = React.useRef<HTMLInputElement>(null);
+
+    const setPage = (allowRetry = false): void => {
+        const newPageNumber = parseInt(inputValue, 10);
+
+        if (
+            !Number.isNaN(newPageNumber) &&
+            newPageNumber >= 1 &&
+            newPageNumber <= pageCount &&
+            newPageNumber !== pageNumber
+        ) {
+            onPageSubmit(newPageNumber);
+        } else {
+            setInputValue(pageNumber.toString()); // Reset the invalid input value to the current page
+
+            if (allowRetry) {
+                setIsButtonFocus(true);
+            }
+        }
+
+        setIsInputShown(false);
+    };
+
+    const handleNumInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+        setInputValue(event.target.value);
+    };
+
+    const handleNumInputBlur = (): void => {
+        setPage();
+    };
+
+    const handleNumInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>): void => {
+        const key = decodeKeydown(event);
+
+        switch (key) {
+            case ENTER:
+                event.stopPropagation();
+                event.preventDefault();
+
+                setPage(true);
+                break;
+            case ESCAPE:
+                event.stopPropagation();
+                event.preventDefault();
+
+                setIsInputShown(false);
+                setIsButtonFocus(true);
+                break;
+            default:
+                break;
+        }
+    };
+
+    React.useEffect(() => {
+        setInputValue(pageNumber.toString()); // Keep internal state in sync with prop as it changes
+    }, [pageNumber]);
+
+    React.useLayoutEffect(() => {
+        if (buttonElRef.current && isButtonFocus) {
+            buttonElRef.current.focus();
+            setIsButtonFocus(false);
```

Comment:
Previously, I had provided an example that had this value stored in a ref and used within a single `useLayoutEffect` hook for `isInputShown`. This approach may trigger unnecessary renders and separates the dependent state. What was the reason for the change?

Reply:
Resolved!  I had an older version of the pr that I was going off of.